File: backend/app/api/v1/endpoints/marcas.py
from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session
from typing import Optional, List
from datetime import date

from app.db.deps import get_marcas_db
from app.services.marcas_service import MarcasService

router = APIRouter()

@router.get("/relojes", response_model=List[str])
def read_relojes(db: Session = Depends(get_marcas_db)):
    """Obtiene la lista de relojes/torniquetes disponibles"""
    service = MarcasService(db)
    return service.get_relojes()

@router.get("/")
def read_marcas(
    limit: int = Query(default=100, ge=1, le=500, description="Cantidad de registros"),
    offset: int = Query(default=0, ge=0, description="Registros a saltar"),
    fecha_inicio: Optional[date] = Query(default=None, description="Fecha inicio del rango (YYYY-MM-DD)"),
    fecha_fin: Optional[date] = Query(default=None, description="Fecha fin del rango (YYYY-MM-DD)"),
    nombre: Optional[str] = Query(default=None, description="Filtrar por nombre (busca en todo el historial)"),
    rut: Optional[str] = Query(default=None, description="Filtrar por RUT (busca en todo el historial)"),
    reloj: Optional[str] = Query(default=None, description="Filtrar por nombre de reloj"),
    tipo_marca: Optional[str] = Query(default=None, description="Filtrar por tipo: IN o OUT"),
    db: Session = Depends(get_marcas_db)
):
    """Obtiene las marcas de empleados con filtros opcionales"""
    service = MarcasService(db)
    marcas, total = service.get_marcas(
        limit, offset, fecha_inicio, fecha_fin, nombre, rut, reloj, tipo_marca
    )
    return {
        "data": marcas,
        "total": total,
        "limit": limit,
        "offset": offset,
        "has_more": offset + len(marcas) < total
    }

# Endpoint legacy para compatibilidad
@router.get("/hoy")
def read_marcas_hoy(
    limit: int = Query(default=100, ge=1, le=500),
    offset: int = Query(default=0, ge=0),
    db: Session = Depends(get_marcas_db)
):
    """Obtiene las marcas del día actual (endpoint legacy)"""
    service = MarcasService(db)
    marcas, total = service.get_marcas(limit, offset)
    return {
        "data": marcas,
        "total": total,
        "limit": limit,
        "offset": offset,
        "has_more": offset + len(marcas) < total
    }

File: frontend/src/services/marcas.js
const API_BASE_URL = 'http://localhost:8000/api/v1';

export const getMarcas = async (params = {}) => {
    const { limit = 100, offset = 0, fechaInicio, fechaFin, nombre, rut, reloj, tipoMarca } = params;
    
    const queryParams = new URLSearchParams();
    queryParams.append('limit', limit);
    queryParams.append('offset', offset);
    
    if (fechaInicio) queryParams.append('fecha_inicio', fechaInicio);
    if (fechaFin) queryParams.append('fecha_fin', fechaFin);
    if (nombre) queryParams.append('nombre', nombre);
    if (rut) queryParams.append('rut', rut);
    if (reloj) queryParams.append('reloj', reloj);
    if (tipoMarca) queryParams.append('tipo_marca', tipoMarca);
    
    const response = await fetch(`${API_BASE_URL}/marcas?${queryParams.toString()}`);
    if (!response.ok) {
        throw new Error('Error al obtener las marcas');
    }
    return response.json();
};

export const getRelojes = async () => {
    const response = await fetch(`${API_BASE_URL}/marcas/relojes`);
    if (!response.ok) {
        throw new Error('Error al obtener los relojes');
    }
    return response.json();
};

// Función legacy para compatibilidad
export const getMarcasHoy = async (limit = 100, offset = 0) => {
    return getMarcas({ limit, offset });
};
